apps/argos_aggregation/lib/argos_aggregation/data_sources/thesauri.ex

```diff
@@ -23,60 +25,215 @@ defmodule ArgosAggregation.Thesauri do
+      url
+      |> HTTPoison.get([], options)
+      |> fetch_response
+    end
+
+    defp fetch_response({:ok, %{status_code: 200, body: body}}), do: {:ok, body}
+    defp fetch_response({:ok, %HTTPoison.Response{status_code: code, request: req}}) do
+      {:error, "Received unhandled status code #{code} for #{req.url}."}
+    end
+    defp fetch_response({:error, error}), do: {:error, error.reason()}
+
+  end
+
+  defmodule DataSourceClient.Local do
+
+    alias ArgosAggregation.ElasticSearch
+
+    def request_single_node(id) do
+      ElasticSearch.DataProvider.get_doc("concept_#{id}")
+    end
+
+  end
+
+
+  defmodule DataProvider do
     require Logger
 
+    alias ArgosAggregation.ElasticSearch.Indexer
+    alias ArgosAggregation.Thesauri.ConceptParser
+
+    def get_by_date(%Date{} = date) do
+      date
+      |> stream_pages
+      |> Stream.map(fn element ->
+        case element do
+          {:error, _err} -> element
+          _ -> ConceptParser.Search.parse_single_doc(element)
+        end
+      end)
+    end
+
+    defp stream_pages(date) do
+      Stream.resource(
+        fn ->
+          case DataSourceClient.Http.request_by_date(date) do
+            {:ok, xml} -> ConceptParser.Search.load_first_page_url(xml)
+            error -> {[error], nil}
+          end
+        end,
+        fn page_url ->
+          case page_url do
+            nil -> {:halt, page_url}
+            page_url -> load_next_page(page_url)
+          end
+        end,
+        fn _val ->
+          Logger.info("Done reading thesaurus updates")
+        end
+      )
+    end
+
+    defp load_next_page(page_url) do
+      case DataSourceClient.Http.read_from_url(page_url) do
+        {:ok, xml} -> ConceptParser.Search.load_next_page_items(xml)
+        error -> {[error], nil}
+      end
+    end
+
+    def get_all() do
+      with {:ok, xml} <- DataSourceClient.Http.request_root_level() do
+        stream_read_hierarchy(xml)
+        |> Stream.flat_map(fn elements ->
+          case elements do
+            {:error, _} -> [elements]
+            _ -> ConceptParser.Hierarchy.read_list_of_descriptions(elements)
+          end
+        end)
+      end
+    end
+
+    defp stream_read_hierarchy(xml) do
+      Stream.resource(
+        fn ->
+          Logger.info("Start reading root level")
+          ConceptParser.Hierarchy.read_root_level(xml)
+        end,
+        #next_fun
+        fn (root_data) ->
+          case root_data do
+            {:error, error} ->
+              Logger.error(error)
+              {[{:error, error}], nil}
+            # last case roots are empty
+            nil ->
+              Logger.info("stopped becaus of nil")
+              {:halt, root_data}
+            [] ->
+              Logger.info("stopped because of empty list")
+              {:halt, root_data}
+            # process roots until all are empty
+            roots ->
+              load_next_nodes(roots)
+          end
+        end,
```

Comment:
Kleinigkeit, aber wäre "node" statt "root" nicht vielleicht der bessere Variablenname (es ist nur beim ersten Durchlauf die root).

Reply:
In dem Fall sind es und bleiben es tatsächlich die Root-Nodes. Der läd am Anfang alle Wurzelkonzepte und läd dann der Reihe nach jeweils deren kompletten Bäume. Die root_data ist dabei die Liste der Wurzelkonzepte, die immer kleiner wird, bis sie leer ist.

Reply:
Ah ok. :+1: